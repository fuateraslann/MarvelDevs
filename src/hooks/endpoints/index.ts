import { useQuery, UseQueryResult } from 'react-query'

import { axios } from 'utils'
import { TCharacter } from 'types'

export const useGetCharacters = (): UseQueryResult<TCharacter[]> => {
  return useQuery(
    'characters',
    async (): Promise<TCharacter[]> => {
      return await axios
        .get(`/v1/public/characters`)
        .then((response) => {
          return response.data.data.results
        })
        .catch((error) => {
          console.log(error)
        })
    },
    {
      refetchOnWindowFocus: false,
    }
  )
}
