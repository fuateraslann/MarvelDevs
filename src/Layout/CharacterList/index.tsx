import { Container, Grid, Typography } from '@mui/material'
import { isEmpty } from 'lodash'
import { CharacterCard } from 'components'
import { useGetCharacters } from 'hooks/endpoints'
import { TCharacter } from 'types'
import { useEffect, useState } from 'react'

const CharacterList = () => {
  const [fetchCharacterOffset, setFetchCharacterOffset] = useState<number>(0)
  const [cardItems, setCardItems] = useState<TCharacter[] | []>([])

  const {
    data: characters,
    isLoading,
    isFetching,
    refetch,
  } = useGetCharacters({
    offset: fetchCharacterOffset,
  })

  const handleScroll = () => {
    const { scrollTop, clientHeight, scrollHeight } = document.documentElement
    if (scrollTop + clientHeight >= scrollHeight - 10 && !isLoading) {
      setFetchCharacterOffset((prevState) => prevState + 30)
    }
  }

  useEffect(() => {
    refetch()
  }, [fetchCharacterOffset])

  useEffect(() => {
    if (characters) {
      setCardItems((prevState) => [...prevState, ...characters])
    }
  }, [characters])

  useEffect(() => {
    window.addEventListener('scroll', handleScroll)
    return () => window.removeEventListener('scroll', handleScroll)
  }, [isLoading])

  if (!cardItems) return null
  if (isEmpty(cardItems) && !isLoading) {
    return <div>No Characters Found ! </div>
  }

  return (
    <Container
      sx={{
        marginTop: '24px',
        height: '100vh',
        width: '100%',
      }}
    >
      <Grid container spacing={2} sx={{ justifyContent: 'center' }}>
        {cardItems?.map((character: TCharacter) => (
          <Grid
            item
            xs={12}
            sm={6}
            md={4}
            sx={{ marginBottom: 7 }}
            key={character.id}
          >
            <CharacterCard character={character} />
          </Grid>
        ))}
        {isFetching && <Typography>LOADING...</Typography>}
      </Grid>
    </Container>
  )
}

export default CharacterList
