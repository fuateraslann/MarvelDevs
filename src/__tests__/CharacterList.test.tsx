import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { ThemeProvider } from '@mui/system'
import { createTheme } from '@mui/material'
import CharacterList from 'Layout/CharacterList'
import { useGetCharacterComics, useGetCharacters } from 'hooks/endpoints'

jest.mock('hooks/endpoints', () => ({
  useGetCharacters: jest.fn(),
  useGetCharacterComics: jest.fn(),
}))

describe('CharacterList', () => {
  const mockCharacters = [
    {
      id: 1,
      name: 'Character 1',
      thumbnail: { path: 'path1', extension: 'jpg' },
    },
    {
      id: 2,
      name: 'Character 2',
      thumbnail: { path: 'path2', extension: 'jpg' },
    },
  ]
  beforeEach(() => {
    (useGetCharacters as jest.Mock).mockReturnValue({
      data: mockCharacters,
      isLoading: false,
      isFetching: false,
      refetch: jest.fn(),
    })
    ;(useGetCharacterComics as jest.Mock).mockReturnValue({
      data: [],
      isLoading: false,
      isFetching: false,
      refetch: jest.fn(),
    })
  })

  test('renders CharacterList correctly', () => {
    const theme = createTheme()

    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <CharacterList />
      </ThemeProvider>
    )

    // Snapshot test
    expect(asFragment()).toMatchSnapshot()
  })
})
