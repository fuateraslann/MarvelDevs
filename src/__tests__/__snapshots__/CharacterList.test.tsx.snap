// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList renders CharacterList correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1nj62pa-MuiContainer-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-1vfpjl-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          style="height: 100%;"
        >
          <img
            alt="Character 1"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            height="90%"
            src="path1.jpg"
          />
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            style="height: 120px; padding: 5px; display: flex; justify-content: space-between;"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
            >
              Character 1
            </h6>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1c8yy18-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Details
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-1vfpjl-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
          style="height: 100%;"
        >
          <img
            alt="Character 2"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
            height="90%"
            src="path2.jpg"
          />
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
            style="height: 120px; padding: 5px; display: flex; justify-content: space-between;"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
            >
              Character 2
            </h6>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1c8yy18-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Details
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
